### PR TITLE
Implement type=

### LINKED ISSUE #16: Write `type=` function
Write a function to check if two `ty`'s represent exactly the same type. Be sure that `(-> (a) a)` matches `(-> (b) b)` but not `(-> (a) b)`.

This can be used to avoid clobbering declared types.

### PR DESCRIPTION
**Pull request by colescott**
Resolves #16

### CODE REVIEW

File: src/types.lisp [hunk 1]
```diff
@@ -124,6 +124,61 @@ If NAME is not known, it will be made known to the global type database."
 
 #+sbcl (declaim (sb-ext:freeze-type ty tyvar tyapp tyfun))
 
+(defun type= (type1 type2)
+  "Check equality of types TYPE1 and TYPE2
+
+Types are equivalent when the structure (TYAPP and TYFUN)
+matches and there exists a bijection between TYVARs of each type."
```
Comment (notmgsk) [inline src/types.lisp]: linebreaks bad

Comment (colescott) [reply src/types.lisp]: My bad. Fixed.

File: src/types.lisp [hunk 2]
```diff
@@ -124,6 +124,60 @@ If NAME is not known, it will be made known to the global type database."
...
+               (let ((pty1 (prune ty1))
+                     (pty2 (prune ty2)))
+                 (cond
+                   ((and (tyvar-p pty1)
+                         (tyvar-p pty2))
+                    (let ((pair1
+                            (find-if
+                             (lambda (x)
+                               (= (tyvar-id pty1)
+                                  (car x)))
```
Comment (stylewarning) [inline src/types.lisp]: `find-if` => `(find (tyvar-id pty1) var-table :key #'car :test #'=)`

File: src/types.lisp [hunk 3]
```diff
@@ -124,6 +124,60 @@ If NAME is not known, it will be made known to the global type database."
...
+                             var-table))
+                          (pair2
+                            (find-if
```
Comment (stylewarning) [inline src/types.lisp]: same here

File: src/types.lisp [hunk 4]
```diff
@@ -124,6 +124,60 @@ If NAME is not known, it will be made known to the global type database."
...
+                        ((equal pair1 pair2)
+                         (when (null pair1)
+                           (push (list (tyvar-id pty1) (tyvar-id pty2))
+                                 var-table))
+                         t)
+                        (t
+                         nil))))
+                   ((and (tyfun-p pty1)
+                         (tyfun-p pty2))
+                    (let ((arity-1 (length (tyfun-from pty1)))
```
Comment (stylewarning) [inline src/types.lisp]: `tyfun-arity` is there for ur taking

Comment (stylewarning) [reply src/types.lisp]: (which means you can collapse the LET and change the last bit to
```
(and (= arity ...) (type= ...) (every ...))
```
without the `if`

File: src/types.lisp [hunk 5]
```diff
@@ -124,6 +124,60 @@ If NAME is not known, it will be made known to the global type database."
...
+                      (if (= arity-1 arity-2)
+                          (and
+                           (%type= (tyfun-to pty1) (tyfun-to pty2))
+                           (every #'%type= (tyfun-from pty1) (tyfun-from pty2)))
+                          nil)))
+                   ((and (tyapp-p pty1)
+                         (tyapp-p pty2))
+                    (let ((name1 (tyapp-name pty1)) (types1 (tyapp-types pty1))
+                          (name2 (tyapp-name pty2)) (types2 (tyapp-types pty2)))
+                      (if (and (eq name1 name2)
```
Comment (stylewarning) [inline src/types.lisp]: collapse the IF into the AND
```
(and (eq name..) (= length ...) (every ...))
```

**Review by stylewarning — CHANGES_REQUESTED** [review]

**Pull request merged**